python/src/cloudstorage/cloudstorage_api.py

```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
   return _Bucket(api, bucket, options)
 
 
+# pylint: disable=too-many-locals, too-many-branches, too-many-statements
+def compose(list_of_files, destination_file, files_metadata=None,
+            content_type=None, retry_params=None, _account_id=None):
+  """
+    Runs the GCS Compose on the given files.
+    Merges between 2 and 32 files into one file. Composite files may even
+      be built from other existing composites, provided that the total
+      component count does not exceed 1024. See here for details:
+      https://cloud.google.com/storage/docs/composite-objects
+  Args:
+    list_of_files: List of file name strings with no leading slashes or bucket.
+    destination_file: Path to the desired output file. Must have the bucket in the path.
+    files_metadata: Optional file metadata, see link for available options:
+                    https://cloud.google.com/storage/docs/composite-objects#_Xml
+    content_type: Used to specify the content-header of the output file.
+    retry_params: An api_utils.RetryParams for this call to GCS. If None,
+    the default one is used.
+    _account_id: Internal-use only.
+
+  Raises:
+    ValueError: If the number of files is outside the range of 2-32.
+  """
+  api = storage_api._get_storage_api(retry_params=retry_params,
+                                   account_id=_account_id)
+  """
+  Needed until cloudstorage_stub.py is updated to accept compose requests
+  TODO: When patched remove the True flow from this if.
+  """
+  if 'development' in os.environ.get('SERVER_SOFTWARE', '').lower():
+    def _temp_func(bucket, file_list, destination_file, content_type):
+      """
+        Dev server stub, remove when the dev server accepts compose requests.
+      """
+      bucket = '/' + destination_file.split('/')[1] + '/'
+      with open(destination_file, 'w', content_type=content_type) as gcs_merge:
+        for source_file in file_list:
+          try:
+            with open(bucket + source_file['Name'], 'r') as gcs_source:
+              gcs_merge.write(gcs_source.read())
+          except errors.NotFoundError as error:
+            logging.warn('File not found %s, skipping', source_file['file_name'])
+            raise error
+    compose_object = _temp_func
+  else:
+    compose_object = api.compose_object
+  file_list, _ = _validate_compose_list(
+                                      destination_file,
+                                      list_of_files,
+                                      files_metadata, 32)
+  compose_object(file_list, destination_file, content_type)
+
+def _compose_batch(list_of_files, destination_file, files_metadata=None,
+            content_type=None, retry_params=None, _account_id=None):
+  """
+  Runs the GCS Compose on the inputed files in batches of 32.
```

Comment:
On Mon, Mar 30, 2015 at 7:25 AM, rbruyere <notifications@github.com> wrote:

> In python/src/cloudstorage/cloudstorage_api.py
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6#discussion_r27394062>
> :
>
> > +          except errors.NotFoundError as error:
> > +            logging.warn('File not found %s, skipping', source_file['file_name'])
> > +            raise error
> > +    compose_object = _temp_func
> > +  else:
> > +    compose_object = api.compose_object
> > +  file_list, _ = _validate_compose_list(
> > +                                      destination_file,
> > +                                      list_of_files,
> > +                                      files_metadata, 32)
> > +  compose_object(file_list, destination_file, content_type)
> > +
> > +def _compose_batch(list_of_files, destination_file, files_metadata=None,
> > +            content_type=None, retry_params=None, _account_id=None):
> > +  """
> > +  Runs the GCS Compose on the inputed files in batches of 32.
>
> I do, not only in personal code but the catalyst for me to start this
> project was to use it in the Mapreduce, which could send up to 256 files
> for the compose.
>
Similar to what we do for Java, such code should be done in the MR (that
could provide the temporary files location and could better handle
cleanup).

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6/files#r27394062>
> .
>